Mathlib/Tactic/GRW.lean

```diff
@@ -0,0 +1,114 @@
+  return ⟨newType, newHyp, subgoals⟩
+
+private partial def _root_.Lean.MVarId.grw (goal : MVarId) (rule : Expr) (rev : Bool := false) :
+    MetaM (MVarId × Array MVarId) := do
+  let ⟨_, prf, mvar, subgoals⟩ ← runGrw (Expr.mvar goal) rule rev true
+  goal.assign prf
+  return ⟨mvar, subgoals⟩
+
+/-- A version of `withRWRulesSeq` (in core) that doesn't attempt to find equation lemmas, and
+allows the caller to maintain state via StateT. -/
+private partial def withRWRulesSeqState {State : Type} (token : Syntax) (rwRulesSeqStx : Syntax)
+    (x : (symm : Bool) → (term : Syntax) → StateT State TacticM Unit) :
+    StateT State TacticM Unit := do
+  let lbrak := rwRulesSeqStx[0]
+  let rules := rwRulesSeqStx[1].getArgs
+  -- show initial state up to (incl.) `[`
+  withTacticInfoContext (mkNullNode #[token, lbrak]) (pure ())
+  let numRules := (rules.size + 1) / 2
+  for i in [:numRules] do
+    let rule := rules[i * 2]!
+    let sep  := rules.getD (i * 2 + 1) Syntax.missing
+    let state ← get
+    -- show rule state up to (incl.) next `,`
+    let newState ← withTacticInfoContext (mkNullNode #[rule, sep]) do
+      -- show errors on rule
+      let s ← withRef rule do
+        let symm := !rule[0].isNone
+        let term := rule[1]
+        -- let processId (id : Syntax) : TacticM Unit := do
+        let ⟨_, newState⟩ ← (x symm term).run state
+        return newState
+      return s
+    set newState
+
+/--
+`grw` is a generalization of the `rw` tactic that takes other relations than equality.  For example,
+```lean
+example (h₁ : a + e ≤ b + e)
+    (h₂ : b < c)
+    (h₃ : c ≤ d) :
+    a + e ≤ d + e := by
+  grw [h₂, h₃] at h₁
+  exact h₁
+```
+
+If applied to a hypothesis or target of type `p` with rule of type `x ~ y` (where `~` is some
+relation) then the resulting type will be `p[x/y]` or the tactic will fail. This tactic will fail
+if side goals are created that it can not fill itself, which it does using `positivity` or
+`assumption`. These two facts make it safe to use in a nonterminal position.
+-/
+elab tok:"grw" rules:rwRuleSeq loc:(location)? : tactic => do
+  withLocation (expandOptLocation (Lean.mkOptionalNode loc))
+    (atLocal := λ fvar => withMainContext do
+      let ⟨_, newGoal, subgoals, _⟩ ← (withRWRulesSeqState tok rules fun rev syn ↦ do
+        let ⟨goal, subgoals, fvar⟩ ← get
+        goal.withContext do
+          let rulePrf ← elabTerm syn none
+          let ⟨newType, newHyp, newSubgoals⟩ ← goal.withContext
+              $ grwHypothesis (Expr.fvar fvar) rulePrf rev
+          let name ← fvar.getUserName
+          let ⟨newFvar, goal', _⟩ ← goal.assertAfter fvar name newType newHyp
+          let newGoal ← goal'.clear fvar
+          set (⟨newGoal, subgoals ++ newSubgoals, newFvar⟩ : MVarId × Array MVarId × FVarId)
+      ).run (⟨← getMainGoal, #[], fvar⟩ : MVarId × Array MVarId × FVarId)
+      let newGoals := subgoals ++ #[newGoal] ++ (← getGoals)
+      setGoals newGoals.toList
+      pruneSolvedGoals
+    )
```

Comment:
I would recommend splitting out the `atLocal` and `atTarget` code into separate declarations (possibly at `MetaM` rather than `TacticM`). It will be more comprehensible, and down the track someone will probably want the MetaM plumbing available for use inside another tactic.